resilience4j-reactor/src/main/java/io/github/resilience4j/reactor/micrometer/operator/TimerSubscriber.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2018 Mariusz Kopylec
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.github.resilience4j.reactor.micrometer.operator;
+
+import io.github.resilience4j.micrometer.Timer;
+import io.github.resilience4j.micrometer.Timer.Context;
+import io.github.resilience4j.reactor.AbstractSubscriber;
+import org.reactivestreams.Subscriber;
+import reactor.core.CoreSubscriber;
+
+import java.lang.reflect.ParameterizedType;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap.KeySetView;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.ConcurrentHashMap.newKeySet;
+
+/**
+ * A Reactor {@link Subscriber} to wrap another subscriber in a timer.
+ *
+ * @param <T> the value type of the upstream and downstream
+ */
+class TimerSubscriber<T> extends AbstractSubscriber<T> {
+
+    private final Context context;
+    private final AtomicBoolean successSignaled = new AtomicBoolean(false);
+    private final boolean singleProducer;
+    private final AtomicReference<T> singleProducerResult = new AtomicReference<>();
+    private final KeySetView<ValueWrapper<T>, Boolean> multipleProducerResult = newKeySet();
+
+    TimerSubscriber(Timer timer, CoreSubscriber<? super T> downstreamSubscriber, boolean singleProducer) {
+        super(downstreamSubscriber);
+        this.singleProducer = singleProducer;
+        context = requireNonNull(timer, "Timer is null").createContext();
+    }
+
+    @Override
+    protected void hookOnNext(T value) {
+        if (!isDisposed()) {
+            if (singleProducer) {
+                singleProducerResult.set(value);
+            } else {
+                multipleProducerResult.add(new ValueWrapper<>(value));
```

Comment:
Concurrent Set is used to collect Flux elements. I'm not sure if this creates any risk of having concurrency problems while processing the reactive stream on a thread pool. I saw that there are Atomic References used all over the project so I decided to use the Concurrent Set.

Those collected elements are passed to Timer's `onResult(...)` method.

Reply:
I think for reactive streams we need to the change how the Timer is working.
The time shouldn't be stopped when a Flux item is emitted, but when the flux is completed successfully or with an error. 
We cannot not capture and store a (possible very long) stream of data in a Set just because of a tag name. 
The questions is how the `onResultTagResolver` should work. 
We could store the first or latest item in a AtomicReference and publish it to `onResult` when the Flux is completed.

Do you have any other smart idea?

Reply:
In our RetryOperator we have a similar issue to handle. 
https://github.com/resilience4j/resilience4j/blob/master/resilience4j-reactor/src/main/java/io/github/resilience4j/reactor/retry/RetryOperator.java#L62

With `upstream.doOnNext(context::handleResult)`  we could store the latest item in some variable.

Reply:
>I think for reactive streams we need to the change how the Timer is working.
The time shouldn't be stopped when a Flux item is emitted, but when the flux is completed successfully or with an error.

This is exactly what I'm trying o achieve here. Every reactive stream (Flux, Flow, Observable, etc) should behave this way. Thus that's how I implemented it. If the implementation is somewhat wrong please let me know.

>The questions is how the onResultTagResolver should work.

Well, that's a hard one. Some options could be:
 - pass first and/or last item emitted
 - pass limited number of first/last items (use Set with limited size)
 - pass nothing
 - pass number of emitted items

Reply:
Sure, that's true. We could also never fire the `onResultTagResolver` for multi-items streams/publishers. We can always fire the `onSuccessTagResolver` for those.

Reply:
Ok, let's try to establish how`onResultTagResolver` should work. Are you OK with any of the options I proposed?

Reply:
Do you have a good usecase for the `onResultTagResolver`.
Otherwise we could also drop it and only provide a `onSuccessTagResolver`

Reply:
At first I thought that it can be used for some business metrics. Suppose you have an inbound/incoming operation like "verify user credentials". Let's say it returns the number of already failed credentials verifications. You assume that if users fail to verify their credentials more than 5 times in row they become suspicious. You would like to measure how many suspicious and not suspicious users are trying to use your system. Now, you can do it either by creating a custom metric or by using `onResultTagResolver` which will check the returned number of already failed credentials verifications.

Now I'm not sure if this the right way to go. I guess these kind of business metrics should be handled separately.
If so, I think we should get remove the `onResultTagResolver` but also the `onSuccessTagResolver`.

Reply:
Yes, let's remove them. Makes it easier to go ahead

Reply:
Ok.
Should I create a separate PR for the removal or just do it here?

Reply:
I would prefer this PR

Reply:
Done